libraries/AC_Avoidance/AP_OADatabase.cpp

```diff
@@ -0,0 +1,467 @@
+    if (_database.size_param != _database.size_param_last && _database.size_param_last != -1) {
+        // param changed, de-allocate to come back and re-allocate later with new size
+        deinit_database();
+        return;
+    } else if (_queue.deinit_needed) {
+        deinit_queue(false);
+        return;
+    } else if (_unable_to_allocate_memory) {
+        return;
+    }
+
+    // try and init and exit immediately if not initialised
+    if (!_initialised) {
+        init();
+        return;
+    }
+
+
+    const uint32_t now_ms = AP_HAL::millis();
+    if (now_ms - _database.optimize_last_ms < 100) {
+        return;
+    }
+
+    _database.optimize_last_ms = now_ms;
+    optimize_db_filter();
+
+    // processing queue by moving those entries into the database
+    // Using a for with fixed size is better than while(!empty) because the
+    // while could get us stuck here longer than expected if we're getting
+    // a lot of values pushing into it while we're trying to empty it. With
+    // the for we know we will exit at an expected time
+    const uint32_t queue_available = _queue.items->available();
+    for (uint32_t queue_index=0; queue_index<queue_available; queue_index++) {
+        OA_DbItem item;
+
+        bool pop_success;
+        {
+            WITH_SEMAPHORE(_queue.sem);
+            pop_success = _queue.items->pop(item);
+        }
+        if (!pop_success) {
+            ::printf("OA: queue pop failure\n");
+            break;
+        }
+
+        // ignore object if close to "latest hit"
+        // the database can reorder as it removed entries, but changes are
+        // good that neighboring entries have similar locations. So, if we find
+        // an item that is good to remove, remember the index and check it first
+        // to try and save ourself from checking the whole database every time
+        bool found = is_item_in_database(_database.last_hit, item);
+
+        // compare item to all items in database. If found a similar item, update the existing, else add it as a new one
+        for (uint32_t i=0; i<_database.count && !found; i++) {
+            // skip last hit because we already checked it above
+            if (i != _database.last_hit && is_item_in_database(i, item)) {
+                found = true;
+                _database.last_hit = i;
+            }
+        }
+
+        // if we got here we should add to database
+        if (!found) {
+            database_item_add(item);
+        } else {
+            _queue.stats.dropped++;
+        }
+    }
+
+    database_items_remove_all_expired();
+}
+
+// returns true if a similar object already exists in database. When true, the object timer is also reset
+bool AP_OADatabase::is_item_in_database(const uint32_t index, const OA_DbItem &item)
+{
+    if (index >= _database.count) {
+        // index out of range
+        return false;
+    }
+    if (_database.items[index].loc.lat == item.loc.lat &&
```

Comment:
The odds of this seem so low why not go directly to the range calculation? If it's an exact match it's still going to pass the distance check, and if it wasn't an exact match this doesn't get you out early. Basically this only saves time when it's a perfect match, which is probably the least likely case, and then doing this while cycling through the full database is going to add up to be expensive.

Reply:
we've removed the check so marking as resolved